File: pylox/pylox/scanner/scanner.py
from .tokens import TokenType, Token
from .. import lox
from ..prelude import *


@dataclass
class Scanner:
	source: str
	position: int = 0
	line: int = 0

	def scan_tokens(self) -> List[Token]:
		tokens: List[Token] = []

		self.consume_empty()
		while self.peek_opt() != None:
			token = self.scan_next()
			if token != None:
				tokens.append(token)
			self.consume_empty()

		tokens.append(
			Token(TokenType.EOF, '', self.line),
		)

		return tokens

	def consume_empty(self):
		while self.position < len(self.source) and self.source[self.position] in [' ', '\t', '\n']:
			if self.source[self.position] == '\n':
				self.line += 1
			self.position += 1

	def peek_opt(self) -> Optional[str]:
		# while self.position < len(self.source) and self.source[self.position] in [' ', '\t', '\n']:
		# 	if self.source[self.position] == '\n':
		# 		self.line += 1
		# 	self.position += 1
		if self.position < len(self.source):
			return self.source[self.position]
	

	def take(self, count: int = 1) -> str:
		rv = self.source[self.position:self.position+count]
		self.position += count
		return rv

	def take_line(self, count: int = 1) -> str:
		index = self.source.find('\n', self.position+1)
		if index == -1:
			index = len(self.source)
		else:
			self.line += 1
			index += 1  # consume the ending /n
		
		rv = self.source[self.position:index]
		self.position = index
		return rv

	def peek_next(self) -> Optional[str]:
		if self.position +1 < len(self.source):
			return self.source[self.position+1]
	
	def get_line_no(self) -> int:
		return 0
	
	def consume_digits(self) -> str:
		rv = ""
		while (a:=self.peek_opt()) != None and a.isnumeric():
			rv += self.take()
		return rv

	def consume_string(self) -> str:
		rv = self.take() # First "
		while (a:=self.peek_opt()) != None and a != '"' :
			if a == "\n":
				self.line += 1
			rv += self.take()
		if self.peek_opt() != '"':
			lox.error(self.line, "Unterminated String")
			return rv
		rv += self.take() # End "
		return rv
	
	def scan_next(self) -> Optional[Token]: # consumes comment if token is comment
		one_char_map = {
			'(': TokenType.LEFT_PARAN,
			')': TokenType.RIGHT_PARAN,
			'{': TokenType.LEFT_BRACE,
			'}': TokenType.RIGHT_BRACE,
			',': TokenType.COMMA,
			'.': TokenType.DOT,
			'-': TokenType.MINUS,
			'+': TokenType.PLUS,
			';': TokenType.SEMICOLON,
			# '/': TokenType.SLASH, # Need to do something related to commenting in lox
			'*': TokenType.STAR
		}


		line_no = self.line

		# Single character
		if self.peek_opt() in one_char_map:
			char = self.take()
			return Token(
				one_char_map[char],
				char,
				line_no
			)

		match self.peek_opt():
			# Either one or two characters (peek next)
			case '!':
				if self.peek_next() == '=':
					return Token(TokenType.BANG_EQUAL, self.take(2), line_no)
				else:
					return Token(TokenType.BANG, self.take(1), line_no)
			
			case '=':
				if self.peek_next() == '=':
					return Token(TokenType.EQUAL_EQUAL, self.take(2), line_no)
				else:
					return Token(TokenType.EQUAL, self.take(1), line_no)
			
			case '>':
				if self.peek_next() == '=':
					return Token(TokenType.GREATER_EQUAL, self.take(2), line_no)
				else:
					return Token(TokenType.GREATER, self.take(1), line_no)

			case '<':
				if self.peek_next() == '=':
					return Token(TokenType.LESS_EQUAL, self.take(2), line_no)
				else:
					return Token(TokenType.LESS, self.take(1), line_no)
			case '/': # 
				if self.peek_next() == '/':
					self.take_line()
					return None
				elif self.peek_next() == '*': # multiline
					self.take()
					self.take()

					while self.peek_next() != None and (self.peek_opt() != '*' or self.peek_next() != '/'):
						self.take()

					if self.peek_next() == None:
						lox.error(line_no, "Unterminated multiline comment")
						return
					self.take() # take *
					self.take() # take /
					return None
				else:
					return Token(TokenType.SLASH, self.take(1), line_no)
				
			case _ as char if char != None and char.isnumeric():
				rv = self.consume_digits()
				if self.peek_opt() == "." and (n:=self.peek_next()) != None and n.isnumeric():
					rv += self.take()
					rv += self.consume_digits()
				return Token(TokenType.NUMBER, rv, line_no, eval(rv))

			case '"':
				rv = self.consume_string()
				return Token(TokenType.STRING, rv, line_no, rv[1:-1])

			case _ as char if char != None and (char.isalpha() or  char == "_"):
				lexeme = self.take()
				while (a:=self.peek_opt()) != None and (a.isalnum() or a == "_"):
					char = self.take()
					lexeme += char

				reserved_identifier = {
					"and": TokenType.AND,
					"class": TokenType.CLASS,
					"else": TokenType.ELSE,
					"false": TokenType.FALSE,
					"true": TokenType.TRUE,
					"fun": TokenType.FUN,
					"for": TokenType.FOR,
					"if": TokenType.IF,
					"nil": TokenType.NIL,
					"or": TokenType.OR,
					"print": TokenType.PRINT,
					"return": TokenType.RETURN,
					"super": TokenType.SUPER,
					"this": TokenType.THIS,
					"var": TokenType.VAR,
					"while": TokenType.WHILE,
					"break": TokenType.BREAK,
				}

				if lexeme in reserved_identifier:
					literal_val = None
					if lexeme == "true":
						literal_val = True
					else:
						literal_val = False
					return Token(reserved_identifier[lexeme], lexeme, line_no, literal_val)
				else:
					return Token(TokenType.IDENTIFIER, lexeme, line_no)

			case _:
				
				lox.error(line_no, "Unexpected token: " + self.take())

File: rlox/src/scanner.rs
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TokenType {
    // Single Character
    /// (
    LeftParen,
    /// )
    RightParen,
    /// {
    LeftBrace,
    /// }
    RightBrace,
    Comma,
    Dot,
    Minus,
    Plus,
    Semicolon,
    Slash,
    Star,

    // one or two character
    Bang,
    BangEqual,
    Equal,
    EqualEqual,
    Greater,
    GreaterEqual,
    Less,
    LessEqual,

    // Literals
    Identifier,
    String,
    Number,

    // Keywords
    And,
    Class,
    Else,
    False,
    For,
    Fun,
    If,
    Nil,
    Or,
    Print,
    Return,
    Super,
    This,
    True,
    Var,
    While,

    // Semantic
    Error,
    Eof,
}

#[derive(Clone, Copy, Debug)]
pub struct Token<'a> {
    pub ty: TokenType,
    pub string: &'a str,
    pub line: usize,
}

pub struct Scanner<'a> {
    src: &'a str,
    start: usize,
    current: usize,
    line: usize,
}

impl<'a> Scanner<'a> {
    pub fn new(source: &'a str) -> Self {
        Self {
            src: source,
            start: 0,
            current: 0,
            line: 1,
        }
    }

    pub fn scan_token(&mut self) -> Token<'a> {
        self.skip_whitespace();
        use TokenType::*;

        self.start = self.current;
        if self.is_at_end() {
            return self.make_token(Eof);
        }

        let c = self.advance();

        match c {
            '(' => return self.make_token(LeftParen),
            ')' => return self.make_token(RightParen),
            '{' => return self.make_token(LeftBrace),
            '}' => return self.make_token(RightBrace),
            ';' => return self.make_token(Semicolon),
            ',' => return self.make_token(Comma),
            '.' => return self.make_token(Dot),
            '-' => return self.make_token(Minus),
            '+' => return self.make_token(Plus),
            '/' => return self.make_token(Slash),
            '*' => return self.make_token(Star),

            '!' => self.conditional_token('=', BangEqual, Bang),
            '=' => self.conditional_token('=', EqualEqual, Equal),
            '<' => self.conditional_token('=', LessEqual, Less),
            '>' => self.conditional_token('=', GreaterEqual, Greater),

            '"' => self.string(),

            c if c.is_numeric() => self.number(),
            c if Self::is_alpha(c) => self.identifier(),

            _ => self.error_token("Unepxected charater"),
        }
    }

    fn advance(&mut self) -> char {
        self.current += 1;

        // TODO(perf): utf-8 indexing will slow things down.
        // If yes, make the content ascii-only.
        // or store the location with string.
        return self.src.chars().nth(self.current - 1).unwrap();
    }

    fn match_char(&mut self, ch: char) -> bool {
        if self.is_at_end() {
            return false;
        }
        if self.src.chars().nth(self.current).unwrap() != ch {
            return false;
        }
        self.current += 1;
        return true;
    }

    fn peek(&self) -> char {
        self.src.chars().nth(self.current).unwrap_or('\0')
    }

    fn peek_next(&self) -> char {
        self.src.chars().nth(self.current + 1).unwrap_or('\0')
    }
    fn is_at_end(&self) -> bool {
        return self.current >= self.src.len();
    }

    fn make_token(&self, ty: TokenType) -> Token<'a> {
        return Token {
            ty,

            // TODO(perf): utf-8 indexing will slow things down.
            // If yes, make the content ascii-only.
            // or store the location with string.
            string: &self.src[self.start..self.current],
            line: self.line,
        };
    }

    #[inline]
    fn conditional_token(&mut self, ch: char, if_yes: TokenType, if_no: TokenType) -> Token<'a> {
        if self.match_char(ch) {
            self.make_token(if_yes)
        } else {
            self.make_token(if_no)
        }
    }

    fn error_token(&self, msg: &'static str) -> Token<'a> {
        return Token {
            ty: TokenType::Error,
            string: msg,
            line: self.line,
        };
    }

    fn skip_whitespace(&mut self) {
        loop {
            let c = self.peek();
            match c {
                ' ' | '\t' | '\r' => {
                    self.advance();
                }
                '\n' => {
                    self.line += 1;
                    self.advance();
                }
                '/' => {
                    if self.peek_next() == '/' {
                        // Rest of the line is a comment
                        while self.peek() != '\n' && !self.is_at_end() {
                            self.advance();
                        }
                    }
                }
                _ => return,
            }
        }
    }

    fn string(&mut self) -> Token<'a> {
        while self.peek() != '"' && !self.is_at_end() {
            if self.peek() == '\n' {
                self.line += 1;
            }
            self.advance();
        }
        if self.is_at_end() {
            return self.error_token("Unterminated String");
        }
        // Consume closing quote
        self.advance();
        return self.make_token(TokenType::String);
    }

    fn number(&mut self) -> Token<'a> {
        while self.peek().is_numeric() {
            self.advance();
        }
        if self.peek() == '.' && self.peek_next().is_numeric() {
            self.advance();
            while self.peek().is_numeric() {
                self.advance();
            }
        }

        self.make_token(TokenType::Number)
    }

    fn identifier(&mut self) -> Token<'a> {
        while Self::is_alpha(self.peek()) || self.peek().is_numeric() {
            self.advance();
        }
        return self.make_token(self.identifier_type());
    }

    fn is_alpha(ch: char) -> bool {
        ch.is_alphabetic() || ch == '_'
    }

    fn check_keyword(&self, val: &str, ty: TokenType) -> TokenType {
        if &self.src[self.start + 1..self.current] == val {
            ty
        } else {
            TokenType::Identifier
        }
    }

    fn identifier_type(&self) -> TokenType {
        use TokenType::*;
        match self.src.chars().nth(self.start).unwrap() {
            'a' => return self.check_keyword("nd", And),
            'c' => return self.check_keyword("lass", Class),
            'e' => return self.check_keyword("lse", Else),
            'f' => {
                if self.current - self.start > 1 {
                    match self.src.chars().nth(self.start + 1).unwrap() {
                        // TODO(perf): unnecessary twice-check
                        'a' => return self.check_keyword("alse", False),
                        'o' => return self.check_keyword("or", For),
                        'u' => return self.check_keyword("un", Fun),
                        _ => (),
                    }
                }
            }
            'i' => return self.check_keyword("f", If),
            'n' => return self.check_keyword("il", Nil),
            'o' => return self.check_keyword("r", Or),
            'p' => return self.check_keyword("rint", Print),
            'r' => return self.check_keyword("eturn", Return),
            's' => return self.check_keyword("uper", Super),
            't' => {
                if self.current - self.start > 1 {
                    match self.src.chars().nth(self.start + 1).unwrap() {
                        // TODO(perf): unnecessary twice-check
                        'h' => return self.check_keyword("his", This),
                        'r' => return self.check_keyword("rue", True),
                        _ => (),
                    }
                }
            }
            'v' => return self.check_keyword("ar", Var),
            'w' => return self.check_keyword("hile", While),
            _ => (),
        };
        return TokenType::Identifier;
    }
}
